### ISSUE TITLE
Make data property an optional generic type

### ISSUE THREAD

**Issue by Avejack** [paperjs/paper.js#1902]
When using paper.js in a Typescript project that has strict linting it can be very annoying to deal with the `data`-property in the `Item`-class to be of type `any`. It would be nice if you could define the type of the `data`-property.

My proposal for this would be to change the `paper.d.ts`-file in the following way:

1. Add an optional generic type to the `Item`-class:
```ts
class Item<T = any> {
```
2. Change type of the `data`-property to the generic type:
```ts
data: T;
```
3. Add an optional generic type to all classes that extend the `Item`-class and get extended themselves and pass the type to the extended class. For example:
```ts
class PathItem<T = any> extends Item<T> {
```

This does **not** make it a breaking change. People who care to change the type of the `data`-property can then instantiate an object and pass their type definition for the `data`-property:

```ts
const item = new paper.Item<{ id: string }>();
item.data // <- now has the `id`-property and gets auto-completed by the IDE.
```

If they don't pass any type it will default to the `any`-type.

I know that you generate the `paper.d.ts`-file from the JSDocs, so you can't just edit the typings file, but maybe you can extend your generation script by this functionality.

This is a very nice package btw!

**Comment by lehni** [comment]
@sasensi this is one for you, I believe. Would be interesting to hear your opinion on this change of the type definitions!

**Comment by sasensi** [comment]
Hi, thank you for your suggestion which is well documented and detailed.
My opinion about it is that I'm not sure that this is such a common use case to add a custom type to the `data` property.
Maybe I could suggest some possible workarounds that are maybe as convenient and sufficient:
```typescript
interface MyData {
    b: number
}

type WithData<Item, Data> = Omit<Item, 'data'> & { data: Data }

function withData<Data, Item>(item: Item) {
    return item as WithData<Item, Data>
}

const item1 = new paper.Item() as WithData<paper.Item, MyData>;
item1.data.b;

const item2 = withData<MyData>(new paper.Item())
item2.data.b;
```
Try them and tell me what you think ?

**Comment by Avejack** [comment]
Yeah, it definitely works that way, but I don't think it's very elegant...

But if you think my suggestion isn't needed enough to change it in the library then I can understand that.

I guess you can close this issue then.
